### PR TITLE
Pthread introspection hook

### PR DESCRIPTION
**Pull request by fabianfreyer**
see #68

### CODE REVIEW

File: libafl_frida/src/lib.rs [hunk 1]
```diff
@@ -10,6 +10,10 @@ pub mod asan_errors;
 /// The frida address sanitizer runtime
 pub mod asan_rt;
 
+/// Hooking thread lifecycle events. Seems like this is apple-only for now.
+#[cfg(target_os = "macos")]
```
Comment (domenukk) [inline libafl_frida/src/lib.rs:14]: Why macos-only?

**Review by domenukk — APPROVED** [review]

File: libafl_frida/src/lib.rs (hunk 1 above)
Comment (fabianfreyer) [reply libafl_frida/src/lib.rs:14]: Only apple implements https://opensource.apple.com/source/libpthread/libpthread-218.20.1/pthread/introspection.h.auto.html

Comment (domenukk) [reply libafl_frida/src/lib.rs:14]: ios too, then?

**Comment by s1341** [comment]
Nice. Didn't know about this api. I would have done it using regular hooks, but this is better I think. Still needs to be wired up so that we start stalking on the new threads...

**Comment by fabianfreyer** [comment]
Well there are basically two things we could do:
- use Frida Interceptor hooks. These work across threads, as they patch the function prologue. It's a bit of a hassle to make these work once a Stalker is active, and decreasing the trustlevel comes at a performance cost.
- use Frida Stalker to emit code on the pthread functions. However, that could end up being a chicken-egg problem, as the stalker needs to be active on the thread that starts the new thread first. When attaching to a multithreaded target, this may be a problem.

I've tried to keep the external API's of the module clean from any apple pthread_introspection internals, so that it would be easy to make a drop-in replacement on other platforms.

**Comment by s1341** [comment]
I think the frida Interceptor hooks are the way to go. I am using them successfully in ASAN. You can see what I did there. There is no need to decrease the trustlevel if you apply the hooks before you start stalking.

**Comment by fabianfreyer** [comment]
Why are those only aarch64 though?

**Comment by fabianfreyer** [comment]
Also, looking at different libc's and their pthread implementations, it isn't really clear to me where to hook for each of them, especially at places where the stack size and location is already known. I doubt function call hooks suffice there; ideally that would be the clone/clone3 syscall on linux. But at least glibc's `__clone_internal` which is called from `pthread_create` doesn't seem to have symbols.

**Comment by s1341** [comment]
FRIDA-ASAN is only aarch64.

Why can we not hook `pthread_create` itself?

**Comment by andreafioraldi** [comment]
Hey @fabianfreyer , what's the state of this PR? libafl_frida is under refactor, you will get some conflicts once we merge the other PRs

**Comment by domenukk** [comment]
Instead of `lazy_static`, maybe take a look at https://crates.io/crates/static_init

**Comment by domenukk** [comment]
Ping @fabianfreyer

File: libafl_frida/src/pthread_hook.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+// At the time where the inner is called, it will have been set.
+// Mark it as sync.
+unsafe impl Sync for PreviousHook {}
+
+#[allow(non_upper_case_globals)]
+static PREVIOUS_HOOK: PreviousHook = PreviousHook(UnsafeCell::new(None));
+
+lazy_static! {
+    static ref CURRENT_HOOK: RwLock<Option<PthreadIntrospectionHook>> = RwLock::new(None);
```
Comment (domenukk) [inline libafl_frida/src/pthread_hook.rs:78]: `RwLock::new` is const, so you can just remove `lazy_static` completely, here

**Comment by domenukk** [comment]
Continuing this in #891

**Pull request closed without merging**